File: src/lib/token-usage.ts
import type {
  TokenUsageBreakdownItem,
  TokenUsageBucket,
  TokenUsageHeatCell,
  TokenUsagePoint,
  TokenUsageReport,
  TokenUsageTotals,
} from "@/lib/types"

/**
 * Pure helpers behind the token-usage dashboard. Everything here is a function
 * of its arguments — no clock, no locale, no DOM — so the range math, the
 * top-N folding and the highlight derivation are exactly testable and the
 * components stay presentational.
 */

// ─── Ranges ─────────────────────────────────────────────────────────────

export type TokenUsageRangePreset =
  | "7d"
  | "30d"
  | "90d"
  | "thisMonth"
  | "thisYear"
  | "all"
  | "custom"

export const RANGE_PRESETS: TokenUsageRangePreset[] = [
  "7d",
  "30d",
  "90d",
  "thisMonth",
  "thisYear",
  "all",
]

export interface ResolvedRange {
  /** ISO-8601, inclusive. `null` = unbounded (the backend clamps to the data). */
  start: string | null
  /** ISO-8601, exclusive. */
  end: string | null
}

/** Midnight at the start of `d`'s local calendar day. */
export function startOfLocalDay(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth(), d.getDate())
}

export function addLocalDays(d: Date, days: number): Date {
  // Constructing through the Date parts (rather than adding milliseconds) keeps
  // the result on a real local midnight across DST transitions.
  return new Date(d.getFullYear(), d.getMonth(), d.getDate() + days)
}

/**
 * The custom range's wire form is a pair of `YYYY-MM-DD` local days; the
 * calendar speaks `Date`. These two are that seam.
 *
 * Neither goes through `toISOString`/`Date.parse` of a bare date: a date-only
 * string is parsed as UTC by the language spec, so west of Greenwich the day
 * would come back as the one before.
 */
export function toLocalDayValue(d: Date): string {
  const pad = (n: number) => String(n).padStart(2, "0")
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`
}

/** `YYYY-MM-DD` → local midnight; `null` for empty or unparsable input. */
export function fromLocalDayValue(value: string): Date | null {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return null
  const parsed = new Date(`${value}T00:00`)
  return Number.isNaN(parsed.getTime()) ? null : parsed
}

/**
 * Which month the two-month range calendar should open on — the LEFT pane, so
 * the right one lands on the range's end (or on the current month when nothing
 * is picked yet).
 *
 * Without this the calendar opens on today and shows [this month, next month],
 * and since there are no statistics in the future that entire right pane is
 * disabled: half the panel is dead and a span reaching back into last month
 * cannot be drawn without first paging backwards. Anchoring one month earlier
 * makes both panes usable and puts a cross-month range within one gesture.
 *
 * Day 1 of the month, because react-day-picker only reads the month off it.
 */
export function rangeDefaultMonth(from: string, to: string, today: Date): Date {
  const anchor = fromLocalDayValue(to) ?? fromLocalDayValue(from) ?? today
  return new Date(anchor.getFullYear(), anchor.getMonth() - 1, 1)
}

/**
 * Turn a preset into concrete UTC bounds.
 *
 * Bounds are always local-day aligned and half-open `[start, end)`: "the last 7
 * days" is the last 7 *calendar* days including today, not a rolling 168 hours,
 * which is what a day-bucketed chart is showing anyway.
 *
 * `custom` takes two inclusive local dates and widens `end` to the following
 * midnight, so picking the same day twice yields that whole day.
 */
export function resolveRange(
  preset: TokenUsageRangePreset,
  now: Date,
  custom?: { from: Date | null; to: Date | null }
): ResolvedRange {
  const tomorrow = addLocalDays(startOfLocalDay(now), 1)
  const iso = (d: Date) => d.toISOString()

  switch (preset) {
    case "7d":
      return { start: iso(addLocalDays(tomorrow, -7)), end: iso(tomorrow) }
    case "30d":
      return { start: iso(addLocalDays(tomorrow, -30)), end: iso(tomorrow) }
    case "90d":
      return { start: iso(addLocalDays(tomorrow, -90)), end: iso(tomorrow) }
    case "thisMonth":
      return {
        start: iso(new Date(now.getFullYear(), now.getMonth(), 1)),
        end: iso(tomorrow),
      }
    case "thisYear":
      return {
        start: iso(new Date(now.getFullYear(), 0, 1)),
        end: iso(tomorrow),
      }
    case "all":
      return { start: null, end: null }
    case "custom": {
      const from = custom?.from ?? null
      const to = custom?.to ?? null
      if (!from && !to) return { start: null, end: null }
      // A half-filled custom range is still meaningful — treat the missing side
      // as unbounded rather than silently falling back to "all".
      return {
        start: from ? iso(startOfLocalDay(from)) : null,
        end: to ? iso(addLocalDays(startOfLocalDay(to), 1)) : null,
      }
    }
  }
}

/**
 * The bucket width that keeps a range readable. Named presets map directly —
 * 90 daily bars is a picket fence, so 90d reads in weeks and a year in
 * months. Custom falls back to the span: days to ~3 months, weeks to a year,
 * months beyond. Only ever used to pick a *default* — the control stays
 * user-settable.
 */
export function suggestBucket(
  preset: TokenUsageRangePreset,
  range: ResolvedRange
): TokenUsageBucket {
  switch (preset) {
    case "7d":
    case "30d":
    case "thisMonth":
      return "day"
    case "90d":
      return "week"
    case "thisYear":
      return "month"
    case "all":
      return "week"
    case "custom":
      break
  }
  if (!range.start || !range.end) return "week"
  const days =
    (new Date(range.end).getTime() - new Date(range.start).getTime()) /
    86_400_000
  if (days <= 92) return "day"
  if (days <= 366) return "week"
  return "month"
}

/** `-getTimezoneOffset()` — minutes east of UTC, the sign the backend expects. */
export function localTzOffsetMinutes(now: Date = new Date()): number {
  return -now.getTimezoneOffset()
}

// ─── Deltas ─────────────────────────────────────────────────────────────

export interface Delta {
  /** Fractional change, e.g. `0.25` for +25%. `null` when there is no base to
   *  compare against (the previous window was empty) — render that as "new",
   *  never as "+∞%" or a misleading "+100%". */
  ratio: number | null
  direction: "up" | "down" | "flat"
}

export function computeDelta(current: number, previous: number): Delta {
  if (previous <= 0) {
    return { ratio: null, direction: current > 0 ? "up" : "flat" }
  }
  const ratio = (current - previous) / previous
  // Sub-half-percent moves read as noise on a spend chart; call them flat so a
  // steady week doesn't sprout arrows.
  if (Math.abs(ratio) < 0.005) return { ratio, direction: "flat" }
  return { ratio, direction: ratio > 0 ? "up" : "down" }
}

// ─── Breakdown folding ──────────────────────────────────────────────────

/**
 * Keep the top `limit` slices and fold the rest into one "other" bucket.
 *
 * The categorical palette has a fixed number of validated slots; a 9th series
 * is never a generated colour. Folding — not cycling — is the rule.
 */
export function foldBreakdown(
  items: TokenUsageBreakdownItem[],
  limit: number
): { shown: TokenUsageBreakdownItem[]; other: TokenUsageBreakdownItem | null } {
  if (items.length <= limit) return { shown: items, other: null }
  const shown = items.slice(0, limit)
  const rest = items.slice(limit)
  const other = rest.reduce<TokenUsageBreakdownItem>(
    (acc, it) => ({
      ...acc,
      input_tokens: acc.input_tokens + it.input_tokens,
      output_tokens: acc.output_tokens + it.output_tokens,
      cache_creation_tokens:
        acc.cache_creation_tokens + it.cache_creation_tokens,
      cache_read_tokens: acc.cache_read_tokens + it.cache_read_tokens,
      total_tokens: acc.total_tokens + it.total_tokens,
      turn_count: acc.turn_count + it.turn_count,
      conversation_count: acc.conversation_count + it.conversation_count,
    }),
    {
      key: "__other__",
      label: "__other__",
      input_tokens: 0,
      output_tokens: 0,
      cache_creation_tokens: 0,
      cache_read_tokens: 0,
      total_tokens: 0,
      turn_count: 0,
      conversation_count: 0,
    }
  )
  return { shown, other: other.total_tokens > 0 ? other : null }
}

// ─── Heatmap ────────────────────────────────────────────────────────────

export interface HeatMatrix {
  /** `[weekday][hour]` totals, 7 × 24, zero-filled. */
  cells: number[][]
  max: number
}

/** Expand the sparse wire cells into the dense grid the chart draws. */
export function buildHeatMatrix(cells: TokenUsageHeatCell[]): HeatMatrix {
  const grid: number[][] = Array.from({ length: 7 }, () =>
    Array.from({ length: 24 }, () => 0)
  )
  let max = 0
  for (const c of cells) {
    if (c.weekday < 0 || c.weekday > 6 || c.hour < 0 || c.hour > 23) continue
    grid[c.weekday][c.hour] += c.total_tokens
    max = Math.max(max, grid[c.weekday][c.hour])
  }
  return { cells: grid, max }
}

// ─── Derived stats ──────────────────────────────────────────────────────

/**
 * Share of input that was served from the prompt cache.
 *
 * The denominator is every token that entered the model as context — fresh
 * input, cache writes, and cache reads — so the ratio answers "how much of what
 * I sent did I not pay full price for". Output tokens are excluded on purpose:
 * they were never cacheable.
 *
 * Counter-shaped rather than totals-shaped so the dashboard (`TokenUsageTotals`)
 * and a single session's live counters (`TurnUsage`, whose fields are named
 * differently) share ONE definition of the ratio instead of each carrying its
 * own division.
 */
export function cacheHitRatio(
  inputTokens: number,
  cacheCreationTokens: number,
  cacheReadTokens: number
): number | null {
  const denominator = inputTokens + cacheCreationTokens + cacheReadTokens
  if (denominator <= 0) return null
  return cacheReadTokens / denominator
}

/** {@link cacheHitRatio} over a dashboard totals row. */
export function cacheHitRate(totals: TokenUsageTotals): number | null {
  return cacheHitRatio(
    totals.input_tokens,
    totals.cache_creation_tokens,
    totals.cache_read_tokens
  )
}

/**
 * Decimal places every surface prints the cache hit rate with
 * ({@link formatPercent}): the dashboard's ring, the share card's copy of it,
 * and the composer popover's row. One number, one reading — a rounded `95%`
 * beside a `94.6%` looks like two different measurements.
 *
 * Deliberately not applied to the archetype copy, whose `{percent}%` sentences
 * are a family of five and read as a badge, not a meter.
 */
export const CACHE_HIT_RATE_DIGITS = 1

export function averagePerConversation(totals: TokenUsageTotals): number {
  if (totals.conversation_count <= 0) return 0
  return totals.total_tokens / totals.conversation_count
}

export function averageTurnsPerConversation(totals: TokenUsageTotals): number {
  if (totals.conversation_count <= 0) return 0
  return totals.turn_count / totals.conversation_count
}

/** Tokens that were actually computed this period — everything the cache did
 *  NOT serve: fresh input, cache writes, and output. */
export function freshTokens(totals: TokenUsageTotals): number {
  return Math.max(0, totals.total_tokens - totals.cache_read_tokens)
}

/**
 * Calendar days inside the report's range with no recorded usage at all.
 *
 * The span is counted in *local calendar days*, matching how the backend
 * counts `active_days` — never as elapsed milliseconds / 86.4M. That matters
 * because an unbounded request comes back clamped to the data: `range_start`
 * is the first activity *instant* (mid-day, not midnight) and `range_end` is
 * one tick past the last, so a ms division would undercount the span by up to
 * two days. `null` when the span is unknowable (no bounds and no activity).
 */
export function idleDays(report: TokenUsageReport): number | null {
  const startIso = report.range_start ?? report.first_activity_at
  const endIso = report.range_end ?? report.last_activity_at
  if (!startIso || !endIso) return null
  let endMs = Date.parse(endIso)
  // `range_end` is exclusive — step back one tick to the last instant that
  // was actually counted. The `last_activity_at` fallback is already
  // inclusive.
  if (report.range_end) endMs -= 1
  const start = new Date(startIso)
  const end = new Date(endMs)
  if (Number.isNaN(start.getTime()) || Number.isNaN(end.getTime())) return null
  const startDay = new Date(
    start.getFullYear(),
    start.getMonth(),
    start.getDate()
  )
  const endDay = new Date(end.getFullYear(), end.getMonth(), end.getDate())
  // Local midnights differ by a whole day count ±1h of DST; rounding
  // absorbs the hour.
  const days =
    Math.round((endDay.getTime() - startDay.getTime()) / 86_400_000) + 1
  if (!Number.isFinite(days) || days <= 0) return null
  return Math.max(0, days - report.totals.active_days)
}

export function averagePerActiveDay(totals: TokenUsageTotals): number {
  if (totals.active_days <= 0) return 0
  return totals.total_tokens / totals.active_days
}

/** The single busiest bucket, or `null` when nothing was recorded. */
export function peakPoint(series: TokenUsagePoint[]): TokenUsagePoint | null {
  let best: TokenUsagePoint | null = null
  for (const p of series) {
    if (p.total_tokens <= 0) continue
    if (!best || p.total_tokens > best.total_tokens) best = p
  }
  return best
}

/** Local hour of day with the most recorded tokens, or `null`. */
export function peakHour(cells: TokenUsageHeatCell[]): number | null {
  const byHour = new Array<number>(24).fill(0)
  for (const c of cells) {
    if (c.hour < 0 || c.hour > 23) continue
    byHour[c.hour] += c.total_tokens
  }
  let best: number | null = null
  for (let h = 0; h < 24; h++) {
    if (byHour[h] <= 0) continue
    if (best === null || byHour[h] > byHour[best]) best = h
  }
  return best
}

// ─── Archetype ──────────────────────────────────────────────────────────

/**
 * A one-line "what kind of builder are you" label for the share card.
 *
 * Every archetype is a threshold on a real measured quantity — nothing is
 * randomised or invented — and the list is ordered most-specific first so the
 * winner is the most interesting true statement about the data, not just the
 * first one that happens to fit. `values` carries the numbers the UI
 * interpolates into the translated string.
 */
export type ArchetypeId =
  | "nightOwl"
  | "earlyBird"
  | "weekendWarrior"
  | "cacheMaster"
  | "marathoner"
  | "polyglot"
  | "laserFocus"
  | "deepDiver"
  | "steady"

export interface Archetype {
  id: ArchetypeId
  /** Interpolation values for the translated label/description. */
  values: Record<string, number>
}

const NIGHT_HOURS = [22, 23, 0, 1, 2, 3]
const MORNING_HOURS = [5, 6, 7, 8]
/** Saturday and Sunday in the backend's Monday-first weekday numbering. */
const WEEKEND_DAYS = [5, 6]

function shareOf(
  cells: TokenUsageHeatCell[],
  predicate: (c: TokenUsageHeatCell) => boolean
): number {
  let total = 0
  let matched = 0
  for (const c of cells) {
    total += c.total_tokens
    if (predicate(c)) matched += c.total_tokens
  }
  return total > 0 ? matched / total : 0
}

export function deriveArchetype(report: TokenUsageReport): Archetype {
  const { totals, heatmap, streak, by_agent, by_folder } = report

  const nightShare = shareOf(heatmap, (c) => NIGHT_HOURS.includes(c.hour))
  if (nightShare >= 0.35) {
    return { id: "nightOwl", values: { percent: Math.round(nightShare * 100) } }
  }

  const morningShare = shareOf(heatmap, (c) => MORNING_HOURS.includes(c.hour))
  if (morningShare >= 0.3) {
    return {
      id: "earlyBird",
      values: { percent: Math.round(morningShare * 100) },
    }
  }

  const weekendShare = shareOf(heatmap, (c) => WEEKEND_DAYS.includes(c.weekday))
  if (weekendShare >= 0.35) {
    return {
      id: "weekendWarrior",
      values: { percent: Math.round(weekendShare * 100) },
    }
  }

  const cache = cacheHitRate(totals)
  if (cache !== null && cache >= 0.7) {
    return { id: "cacheMaster", values: { percent: Math.round(cache * 100) } }
  }

  if (streak.longest_days >= 7) {
    return { id: "marathoner", values: { days: streak.longest_days } }
  }

  const significantAgents = by_agent.filter(
    (a) =>
      totals.total_tokens > 0 && a.total_tokens / totals.total_tokens >= 0.1
  ).length
  if (significantAgents >= 3) {
    return { id: "polyglot", values: { count: significantAgents } }
  }

  const topFolder = by_folder[0]
  if (
    topFolder &&
    totals.total_tokens > 0 &&
    topFolder.total_tokens / totals.total_tokens >= 0.7 &&
    by_folder.length > 1
  ) {
    return {
      id: "laserFocus",
      values: {
        percent: Math.round(
          (topFolder.total_tokens / totals.total_tokens) * 100
        ),
      },
    }
  }

  const perConversation = averagePerConversation(totals)
  if (perConversation >= 500_000) {
    // Thousands, so the translated string can interpolate a plain number
    // ("{averageK}K tokens") without needing a formatter injected here.
    return {
      id: "deepDiver",
      values: { averageK: Math.round(perConversation / 1000) },
    }
  }

  return { id: "steady", values: { days: totals.active_days } }
}

// ─── Formatting ─────────────────────────────────────────────────────────

/**
 * Token counts for headline positions: two significant decimals so a share
 * card reads "1.28M" rather than the lossy "1.3M" the compact list formatter
 * produces.
 */
export function formatTokensPrecise(n: number): string {
  if (!Number.isFinite(n) || n < 0) return "0"
  if (n >= 1_000_000_000) return `${(n / 1_000_000_000).toFixed(2)}B`
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(2)}M`
  if (n >= 10_000) return `${(n / 1_000).toFixed(1)}K`
  return Math.round(n).toLocaleString()
}

/** `1h 24m` / `3m 05s` / `12s` — generation time, never zero-padded hours. */
export function formatDuration(ms: number): string {
  if (!Number.isFinite(ms) || ms <= 0) return "0s"
  const totalSeconds = Math.floor(ms / 1000)
  const hours = Math.floor(totalSeconds / 3600)
  const minutes = Math.floor((totalSeconds % 3600) / 60)
  const seconds = totalSeconds % 60
  if (hours > 0) return `${hours}h ${minutes}m`
  if (minutes > 0) return `${minutes}m ${String(seconds).padStart(2, "0")}s`
  return `${seconds}s`
}

export function formatPercent(ratio: number, digits = 0): string {
  return `${(ratio * 100).toFixed(digits)}%`
}

/** A filename-safe stamp for the exported share card. */
export function shareCardFilename(range: ResolvedRange, now: Date): string {
  const day = (iso: string) => iso.slice(0, 10)
  const span =
    range.start && range.end
      ? `${day(range.start)}_${day(new Date(new Date(range.end).getTime() - 1).toISOString())}`
      : "all-time"
  return `codeg-tokens-${span}-${day(now.toISOString())}.png`
}

File: src-tauri/src/models/token_usage.rs
//! Wire types for the token-usage dashboard.
//!
//! ## What "usage" means here
//!
//! One row of [`crate::db::entities::token_usage_turn`] is one *API call's*
//! usage as the agent's own transcript reported it. Summing them is the same
//! semantic `parsers::compute_session_stats` already uses for a single
//! conversation's total — input tokens legitimately repeat across turns
//! because every call re-sends the conversation prefix, and that repetition is
//! exactly what a spend dashboard is measuring.
//!
//! One nuance worth stating: for Codex, `SessionStats::total_usage` shown in
//! the conversation panel is the agent's own authoritative cumulative counter,
//! not the sum of turns. The dashboard always sums turns (it has to — a
//! session-level total cannot be placed on a time axis), so a Codex
//! conversation's dashboard total can differ slightly from the number its
//! detail panel prints.
//!
//! ## Time
//!
//! Facts are stored in UTC. Every bucket boundary (day / week / month / hour of
//! day / weekday) is computed in the *viewer's* local time, using the
//! `tz_offset_minutes` the client sends. A fixed offset is used for the whole
//! range rather than a real tz database: a DST transition inside the range
//! shifts one boundary by an hour, which is immaterial at this granularity and
//! keeps the aggregation a pure, testable function.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Bucket width of the time series.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum TokenUsageBucket {
    #[default]
    Day,
    /// ISO weeks — Monday-start. Not configurable: every locale dextra ships
    /// either uses Monday or reads a Monday-start week without friction, and a
    /// switch would have to thread through the streak math too.
    Week,
    Month,
}

/// What the dashboard is asking for.
///
/// `None` on a dimension means "no constraint" — every value passes. An empty
/// `Vec` is NOT the same thing: it is an explicit selection of nothing, and
/// matches nothing. The frontend sends `null` (never `[]`) when a filter is
/// cleared, so the two never blur in practice; the distinction exists so a
/// caller that really means "nothing" can say so instead of silently getting
/// everything.
#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TokenUsageFilter {
    /// Inclusive lower bound. `None` = from the first recorded turn.
    pub start: Option<DateTime<Utc>>,
    /// Exclusive upper bound. `None` = up to now.
    pub end: Option<DateTime<Utc>>,
    /// Selected project folders. Each id is expanded to include the folders
    /// whose `parent_id` points at it, so picking a project also counts the
    /// task worktrees created under it.
    pub folder_ids: Option<Vec<i32>>,
    /// `conversation.agent_type` wire strings (`"claude_code"`, `"custom:x"`).
    pub agent_types: Option<Vec<String>>,
    /// Exact model names as the transcripts reported them.
    pub models: Option<Vec<String>>,
    #[serde(default)]
    pub bucket: TokenUsageBucket,
    /// The viewer's offset from UTC in minutes (`-new Date().getTimezoneOffset()`).
    #[serde(default)]
    pub tz_offset_minutes: i32,
    /// Also compute totals for the equally-long window immediately before
    /// `start`, so the UI can show period-over-period deltas. Ignored when the
    /// range is open-ended on either side.
    #[serde(default)]
    pub compare_previous: bool,
}

/// The four raw counters plus everything derivable from a set of turns.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TokenUsageTotals {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_creation_tokens: u64,
    pub cache_read_tokens: u64,
    pub total_tokens: u64,
    pub turn_count: u64,
    /// In a served report this is the *workspace* session count for the window
    /// (see `workspace_conversation_count`), overwritten by the command layer
    /// so it reconciles with the status bar's counter. The fold itself fills
    /// in distinct fact conversations, which is what breakdown items and
    /// `previous_totals` keep.
    pub conversation_count: u64,
    /// Summed generation time of the counted turns — not wall-clock time spent
    /// in the app.
    pub duration_ms: u64,
    /// Distinct local calendar days with at least one counted turn.
    pub active_days: u32,
}

/// One point of the time series. `bucket_key` is a stable local-time label the
/// frontend formats for display; `start`/`end` are the real UTC bounds so a
/// tooltip can render the exact span.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsagePoint {
    /// `YYYY-MM-DD` for day and week buckets (the bucket's first local day),
    /// `YYYY-MM` for month buckets.
    pub bucket_key: String,
    pub start: DateTime<Utc>,
    pub end: DateTime<Utc>,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_creation_tokens: u64,
    pub cache_read_tokens: u64,
    pub total_tokens: u64,
    pub turn_count: u64,
    pub conversation_count: u64,
}

/// One slice of a categorical breakdown (folder / agent / model).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageBreakdownItem {
    /// Stable identity for color assignment and filter round-trips: the folder
    /// id as a string, the agent wire name, or the model name.
    pub key: String,
    /// Display name resolved server-side (folder alias/name; model and agent
    /// keys are their own label).
    pub label: String,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_creation_tokens: u64,
    pub cache_read_tokens: u64,
    pub total_tokens: u64,
    pub turn_count: u64,
    pub conversation_count: u64,
}

/// One cell of the weekday × hour activity grid. Only non-empty cells are sent.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageHeatCell {
    /// 0 = Monday … 6 = Sunday, in local time.
    pub weekday: u8,
    /// 0–23, local time.
    pub hour: u8,
    pub total_tokens: u64,
    pub turn_count: u64,
}

/// A single conversation's contribution, for the "biggest sessions" list.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageConversationItem {
    pub conversation_id: i32,
    pub title: Option<String>,
    pub agent_type: String,
    pub folder_label: Option<String>,
    pub total_tokens: u64,
    pub turn_count: u64,
    pub last_activity_at: DateTime<Utc>,
}

/// Consecutive-active-day runs over local calendar days inside the range.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TokenUsageStreak {
    /// Longest run of consecutive active days anywhere in the range.
    pub longest_days: u32,
    /// Run ending on the range's last active day — only meaningful when that
    /// day is today or yesterday, which the frontend decides.
    pub current_days: u32,
    /// Local date (`YYYY-MM-DD`) the current run ends on.
    pub current_ends_on: Option<String>,
}

/// Everything one dashboard render needs, from one query.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageReport {
    /// The range actually applied after resolving open bounds — what the share
    /// card prints.
    pub range_start: Option<DateTime<Utc>>,
    pub range_end: Option<DateTime<Utc>>,
    pub bucket: TokenUsageBucket,
    pub totals: TokenUsageTotals,
    /// Totals for the equally-long window before `range_start`. `None` unless
    /// `compare_previous` was set and both bounds resolved.
    pub previous_totals: Option<TokenUsageTotals>,
    /// Dense: buckets with no activity inside the range are present with zeros,
    /// so the chart's x-axis has no invisible gaps.
    pub series: Vec<TokenUsagePoint>,
    pub by_folder: Vec<TokenUsageBreakdownItem>,
    pub by_agent: Vec<TokenUsageBreakdownItem>,
    pub by_model: Vec<TokenUsageBreakdownItem>,
    pub heatmap: Vec<TokenUsageHeatCell>,
    pub top_conversations: Vec<TokenUsageConversationItem>,
    pub streak: TokenUsageStreak,
    pub first_activity_at: Option<DateTime<Utc>>,
    pub last_activity_at: Option<DateTime<Utc>>,
    /// True when the query hit [`crate::commands::token_usage::MAX_SCANNED_FACTS`]
    /// and the numbers are a prefix of the real range rather than all of it.
    pub truncated: bool,
}

/// A folder that has at least one recorded turn.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageFolderFacet {
    pub folder_id: i32,
    /// Compact display name — the alias when one is set, else the folder name.
    /// Kept for callers that only have room for one string; the filter list
    /// renders `alias [ name ]` from the two fields below instead, so an
    /// aliased folder never hides which directory it actually is.
    pub label: String,
    /// The folder's real (on-disk) directory name, alias or not.
    pub name: String,
    /// The user-set display alias, or `None` when unset.
    pub alias: Option<String>,
    pub path: String,
    /// Set when this folder is a worktree created under another folder — the
    /// UI nests it so picking the parent visibly covers it.
    pub parent_id: Option<i32>,
}

/// The dimension values that actually occur in the stored facts, so the filter
/// controls never offer an option that yields an empty chart.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageFacets {
    pub folders: Vec<TokenUsageFolderFacet>,
    /// Agent wire names, ordered by recorded tokens descending.
    pub agents: Vec<String>,
    /// Model names, ordered by recorded tokens descending.
    pub models: Vec<String>,
    pub data_start: Option<DateTime<Utc>>,
    pub data_end: Option<DateTime<Utc>>,
}

/// How current the materialized facts are.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenUsageSyncStatus {
    /// Live (non-deleted) conversations that can carry usage.
    pub total_conversations: u64,
    /// Of those, how many have been parsed at least once.
    pub synced_conversations: u64,
    /// Of those, how many changed since their last parse (includes the
    /// never-parsed).
    pub stale_conversations: u64,
    pub fact_rows: u64,
    pub last_synced_at: Option<DateTime<Utc>>,
    /// A sync is running right now (this process).
    pub running: bool,
}

/// Outcome of one sync pass.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TokenUsageSyncResult {
    pub scanned: u32,
    /// Conversations re-parsed and rewritten.
    pub synced: u32,
    /// Conversations skipped because nothing changed since the last pass.
    pub skipped: u32,
    /// Conversations that hit a real fault — a transcript that would not parse,
    /// a write that errored. Their previous facts are left untouched and the
    /// next pass retries them. This is the only counter the UI treats as bad
    /// news, so it must never carry a state the user cannot act on.
    pub failed: u32,
    /// Conversations whose transcript is gone. Their recorded facts are kept
    /// exactly as they were and their stamp is settled, so a source that can
    /// never be re-derived stops being re-attempted (and re-reported) on every
    /// pass. Not a failure: nothing was lost that was not already lost on disk.
    pub lost: u32,
    pub turns_written: u64,
    pub tokens_written: u64,
    /// Fact rows dropped because their conversation no longer exists.
    pub pruned_conversations: u32,
}

/// Progress ticks emitted while a sync runs.
#[derive(Debug, Clone, Serialize)]
pub struct TokenUsageSyncProgress {
    pub done: u32,
    pub total: u32,
    /// Title of the conversation just finished — a human-readable "what is it
    /// doing" line, absent for untitled sessions.
    pub current_title: Option<String>,
    /// Set on the final tick; absent while running.
    pub result: Option<TokenUsageSyncResult>,
}
